### PR TITLE
TMA-1078: Generate stronger passwords

### PR DESCRIPTION
**Pull request by panjan**
Introduces usage of the secure SecureRandom.base64
method for generating passwords. Adds extra characters
to the password to ensure presence of 4 character groups
(lower case, upper case, digit, symbol).

### CODE REVIEW

**Comment by gdgate** [comment]
Build in pipeline check aborted.

File: lib/gooddata/helpers/crypto_helper.rb [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        private
+
+        # Pseudo-randomly "sprinkles" the given string
+        # with 4 character groups (digit, lower case,
+        # upper case, symbols).
+        # @param [String] password
+        def sprinkle(password)
+          password_dup = password.dup
+          positions = 0..password.size
+          password_dup.insert(rand(positions), digit)
```
Comment (yenkins) [inline lib/gooddata/helpers/crypto_helper.rb]: info: Calls 'rand(positions)' 4 times (DuplicateMethodCall) (reek)

File: lib/gooddata/helpers/crypto_helper.rb [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+#
+# Copyright (c) 2010-2018 GoodData Corporation. All rights reserved.
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+require 'securerandom'
+
+module GoodData
+  module Helpers
+    module CryptoHelper
```
Comment (yenkins) [inline lib/gooddata/helpers/crypto_helper.rb]: info: Has no descriptive comment (IrresponsibleModule) (reek)

File: lib/gooddata/helpers/crypto_helper.rb [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+          sprinkle(SecureRandom.base64(32))
+        end
+
+        private
+
+        # Pseudo-randomly "sprinkles" the given string
+        # with 4 character groups (digit, lower case,
+        # upper case, symbols).
+        # @param [String] password
+        def sprinkle(password)
```
Comment (yenkins) [inline lib/gooddata/helpers/crypto_helper.rb]: info: Has approx 7 statements (TooManyStatements) (reek)

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/613/ : SUCCESS in 10s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1436/ : SUCCESS in 3m 28s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1704/ : SUCCESS in 1m 48s

**Comment by panjan** [comment]
extended test

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/614/ : SUCCESS in 7s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1530/ : SUCCESS in 35m 40s
- https://checklist.intgdc.com/job/lcm/job/gooddata-ruby-lcm-integration-tests-zuul/98/ : SUCCESS in 25m 54s

**Comment by panjan** [comment]
The code under that link relies on the `#sample` method which is not cryptographically secure. I believe that one should not write his own cryptographic code. Therefore, I chose to use the `SecureRandom.base64` method which uses a secure random number generator. The `.sprinkle` method is just an addition to fulfil our special password requirements.

**Comment by panjan** [comment]
For the record, we've agreed to merge this PR now and keep working on a better solution for generating passwords on the server side.

**Pull request merged**